apollo-router/src/batching.rs

```diff
@@ -0,0 +1,694 @@
+//! Various utility functions and core structures used to implement batching support within
+//! the router.
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt;
+use std::sync::Arc;
+
+use hyper::Body;
+use opentelemetry::trace::TraceContextExt;
+use opentelemetry::Context as otelContext;
+use parking_lot::Mutex;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+use tower::BoxError;
+use tracing::Instrument;
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+use crate::error::FetchError;
+use crate::error::SubgraphBatchingError;
+use crate::graphql;
+use crate::query_planner::fetch::QueryHash;
+use crate::services::http::HttpClientServiceFactory;
+use crate::services::process_batches;
+use crate::services::SubgraphRequest;
+use crate::services::SubgraphResponse;
+use crate::Context;
+
+/// A query that is part of a batch.
+///
+/// Note: We do NOT want this to implement `Clone` because it holds a sender
+/// to the batch, which is waiting for all batch queries to drop their senders
+/// in order to finish processing the batch.
+#[derive(Debug)]
+pub(crate) struct BatchQuery {
+    /// The index of this query relative to the entire batch
+    index: usize,
+
+    /// A channel sender for sending updates to the entire batch
+    sender: Option<mpsc::Sender<BatchHandlerMessage>>,
+
+    /// How many more progress updates are we expecting to send?
+    remaining: usize,
+
+    /// Batch to which this BatchQuery belongs
+    batch: Arc<Batch>,
+}
```

Comment:
could this structure be made Clone, Sync and Send? That would avoid the frequent danse of removing it from extensions then putting it back

Reply:
I've raised a PR against this PR to explore this: https://github.com/apollographql/router/pull/4941

Please let me know what you think about that.

Reply:
done: 86855eebe883f211274a51d9449eeea5e51a0aaf